<h1>
  <span style="color:#0d8dfc;">ACTIVITY 2:</span>
  <span>UK ACCIDENT DATA ANALYTICS PROJECT</span>
</h1>

<hr>

<h1> ANALYST: RHEALYN VASQUEZ </h1>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.stats import f_oneway
import warnings
warnings.filterwarnings('ignore')

In [2]:
uk_accidents = pd.read_csv('datasets\\uk_accident_data.csv')

In [3]:
#this shows the 5 first row
uk_accidents.head()

Index Accident_Severity Accident Date   Latitude  \
0  200701BS64157           Serious    05/06/2019  51.506187   
1  200701BS65737           Serious    02/07/2019  51.495029   
2  200701BS66127           Serious    26/08/2019  51.517715   
3  200701BS66128           Serious    16/08/2019  51.495478   
4  200701BS66837            Slight    03/09/2019  51.488576   

              Light_Conditions           District Area  Longitude  \
0        Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                     Daylight  Kensington and Chelsea  -0.173647   
2  Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                     Daylight  Kensington and Chelsea  -0.202731   
4        Darkness - lights lit  Kensington and Chelsea  -0.192487   

   Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                     1                   2                     Dry   
1                     1                   2             Wet or damp   
2                     1                   3                     Dry   
3                     1                   4                     Dry   
4                     1                   2                     Dry   

            Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0  Single carriageway               Urban     Fine no high winds   
1  Single carriageway               Urban  Raining no high winds   
2                 NaN               Urban                    NaN   
3  Single carriageway               Urban     Fine no high winds   
4                 NaN               Urban                    NaN   

                           Vehicle_Type  
0                                   Car  
1                                   Car  
2                 Taxi/Private hire car  
3  Bus or coach (17 or more pass seats)  
4                         Other vehicle

In [4]:
#this shows quick summary of the DataFrame
uk_accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [5]:
# this shows quick statistics about the data
uk_accidents.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [6]:
uk_accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

<h1> FILLING MISSING VALUES/NULL VALUES </h1>

In [7]:
#mean - to handle missing values in numeric columns
for num in ['Latitude', 'Longitude']:
    uk_accidents[num] = uk_accidents[num].fillna(uk_accidents[num].mean())
    
#mode - to handle missing values in categorical columns
for cat in ['Road_Surface_Conditions', 'Road_Type', 'Urban_or_Rural_Area', 'Weather_Conditions']:
    uk_accidents[cat] = uk_accidents[cat].fillna(uk_accidents[cat].mode()[0])


In [8]:
uk_accidents.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [9]:
uk_accidents['Index'] = uk_accidents['Index'].astype('category')
uk_accidents['Accident_Severity'] = uk_accidents['Accident_Severity'].astype('category')
uk_accidents['Accident Date'] = pd.to_datetime(uk_accidents['Accident Date'], errors='coerce')
uk_accidents['Latitude'] = uk_accidents['Latitude'].astype('float')
uk_accidents['Longitude'] = uk_accidents['Longitude'].astype('float')
uk_accidents['Light_Conditions'] = uk_accidents['Light_Conditions'].astype('category')
uk_accidents['District Area'] = uk_accidents['District Area'].astype('category')
uk_accidents['Number_of_Casualties'] = uk_accidents['Number_of_Casualties'].astype('int')
uk_accidents['Number_of_Vehicles'] = uk_accidents['Number_of_Vehicles'].astype('int')
uk_accidents['Road_Surface_Conditions'] = uk_accidents['Road_Surface_Conditions'].astype('category')
uk_accidents['Road_Type'] = uk_accidents['Road_Type'].astype('category')
uk_accidents['Urban_or_Rural_Area'] = uk_accidents['Urban_or_Rural_Area'].astype('category')
uk_accidents['Weather_Conditions'] = uk_accidents['Weather_Conditions'].astype('category')
uk_accidents['Vehicle_Type'] = uk_accidents['Vehicle_Type'].astype('category')

In [10]:
uk_accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   Index                    660679 non-null  category      
 1   Accident_Severity        660679 non-null  category      
 2   Accident Date            265007 non-null  datetime64[ns]
 3   Latitude                 660679 non-null  float64       
 4   Light_Conditions         660679 non-null  category      
 5   District Area            660679 non-null  category      
 6   Longitude                660679 non-null  float64       
 7   Number_of_Casualties     660679 non-null  int64         
 8   Number_of_Vehicles       660679 non-null  int64         
 9   Road_Surface_Conditions  660679 non-null  category      
 10  Road_Type                660679 non-null  category      
 11  Urban_or_Rural_Area      660679 non-null  category      
 12  Weather_Conditio

<h1 style="color:#0d8dfc;"> DATA ANALYSIS</h1>
<hr>

<h1> DATA AGGREGATION</h1>

<h1> 1. Which district has the highest average number of casualties per accident? </h1>

In [11]:
highest_casualties = uk_accidents.groupby("District Area")["Number_of_Casualties"].mean().sort_values(ascending=False).head(1)
highest_casualties

District Area
Blaeu Gwent    1.693694
Name: Number_of_Casualties, dtype: float64

<h1> INSIGHT 1:</h1>
<p> This shows that the district with the highest average casualties per accident is Blaeu Gwent, where each crash injures about 1.69 people on average.</p>

<h1> 2. Which road type has the highest average number of vehicles involved? </h1>

In [12]:
uk_accidents.groupby("Road_Type")["Number_of_Vehicles"].mean().sort_values(ascending=False)

Road_Type
Dual carriageway      2.009766
Slip road             1.917483
Roundabout            1.901323
Single carriageway    1.794803
One way street        1.585441
Name: Number_of_Vehicles, dtype: float64

<h1> INSIGHT 2: </h1>
<p> This shows that crashes on dual carriageways usually involve the most vehicles, while one-way streets involve the fewest. </p>

<h1> 3. Which weather condition is linked with the highest average accident severity? </h1>

In [13]:
uk_accidents['Accident_Severity'] = uk_accidents['Accident_Severity'].cat.codes + 1
uk_accidents.groupby("Weather_Conditions")['Accident_Severity'].mean().sort_values(ascending=False)

Weather_Conditions
Snowing no high winds    2.897884
Other                    2.875743
Snowing + high winds     2.870056
Raining no high winds    2.859918
Raining + high winds     2.838690
Fine no high winds       2.836081
Fog or mist              2.816610
Fine + high winds        2.813538
Name: Accident_Severity, dtype: float64

<h1> INSIGHT 3: </h1>
<p> This shows that the accidents during snowy conditions or unusual weather “Other” tend to be more severe on average, while fine or foggy weather sees slightly lower severity.</p>

<h1> 4. How does the average number of casualties from accidents differ between urban and rural areas? </h1>

In [14]:
uk_accidents.groupby("Urban_or_Rural_Area")["Number_of_Casualties"].mean().sort_values(ascending=False)

Urban_or_Rural_Area
Rural          1.479204
Urban          1.287807
Unallocated    1.181818
Name: Number_of_Casualties, dtype: float64

<h1> INSIGHT 4: </h1>
<p> This shows that the accidents in rural areas have a higher average number of casualties (1.48) compared to urban areas (1.29), with unallocated areas having the lowest average (1.18). This suggests that rural accidents tend to be more severe, possibly due to higher speeds or less safe road conditions.
</p>

<h1> 5. In which light conditions do accidents typically have the highest severity?</h1>

In [15]:
uk_accidents.groupby("Light_Conditions")["Accident_Severity"].mean().sort_values(ascending=False)

Light_Conditions
Darkness - lighting unknown    2.856570
Daylight                       2.853756
Darkness - lights lit          2.823327
Darkness - lights unlit        2.823044
Darkness - no lighting         2.722253
Name: Accident_Severity, dtype: float64

<h1> INSIGHT 5:</h1>
<p>This shows that while darkness with unknown lighting(2.856570 or2.86) is linked to the most severe accidents on average, severity differences between lighting conditions are relatively small.</p>

<hr> 
<h1> Correlation </h1>

<h1> 6. How is the number of casualties related to the number of vehicles involved in an accident? </h1>

In [16]:
uk_accidents["Number_of_Casualties"].corr(uk_accidents["Number_of_Vehicles"])

np.float64(0.2288888612692756)

<h1> INSIGHT 6: </h1>
<p>A positive correlation of 0.23 indicates that accidents with more vehicles tend to have slightly more casualties, though the relationship is weak.
</p>

<h1>7. Does the latitude affect how severe accidents are? </h1>

In [17]:
uk_accidents["Latitude"].corr(uk_accidents["Accident_Severity"])

np.float64(-0.026468995471298485)

<h1>INSIGHT 7:</h1>
<p>This shows that there is no correlation between latitude and accident severity, meaning that where an accident occurs north or south does not affect how severe it is.
</p>

<h1> 8. Has the number of casualties changed over time?</h1>

In [18]:
uk_accidents["Accident Date"] = pd.to_datetime(uk_accidents["Accident Date"], errors='coerce')
uk_accidents = uk_accidents.dropna(subset=["Accident Date"])
uk_accidents["Date_num"] = uk_accidents["Accident Date"].map(pd.Timestamp.toordinal)
uk_accidents["Date_num"].corr(uk_accidents["Number_of_Casualties"])

np.float64(-0.003031562465276751)

<h1> INSIGHT 8:</h1>
<p>This shows that there is no correlation between accident date and the number of casualties, indicating that the number of casualties has not changed significantly over time.
</p>

<hr>
<h1>DESCRIPTIVE</h1>

<h1>9. Which types of roads are most commonly involved in accidents? </h1>

In [19]:
road_counts = uk_accidents["Road_Type"].value_counts().sort_values(ascending=False)
road_counts

Road_Type
Single carriageway    199555
Dual carriageway       39632
Roundabout             17530
One way street          5468
Slip road               2822
Name: count, dtype: int64

<h1> INSIGHT 9: </h1>
<p> This suggests that single carriageways are the most accident-prone, possibly due to higher traffic density, limited overtaking opportunities, or narrower lanes.</p>

<h1> 10. How are accidents distributed between urban and rural areas?</h1>

In [20]:
area_counts = uk_accidents["Urban_or_Rural_Area"].value_counts()
area_counts

Urban_or_Rural_Area
Urban          169042
Rural           95961
Unallocated         4
Name: count, dtype: int64

<h1>INSIGHT 10: </h1>
<p>This indicates that urban environments, likely due to higher traffic density and more complex road networks, experience more accidents than rural areas.</p>

<hr>
<h1> ANOVA (Categorical → Numerical) </h1>

<h1>11. Do different road types have different average casualties?</h1>

In [21]:
groups = [g["Number_of_Casualties"].values for _, g in uk_accidents.groupby("Road_Type")]
f_oneway(*groups)

F_onewayResult(statistic=np.float64(298.0796466734802), pvalue=np.float64(2.793590455313895e-256))

<h1>INSIGHT 11:</h1>
<p>This shows that the average number of casualties differs significantly between road types, with some roads, like single carriageways, tending to have more casualties than others.
</p>

<h1> 12.  Does light condition affect the number of vehicles involved? </h1>

In [22]:
groups = [g["Number_of_Vehicles"].values for _, g in uk_accidents.groupby("Light_Conditions")]
f_oneway(*groups)

F_onewayResult(statistic=np.float64(595.5530377387355), pvalue=np.float64(0.0))

<h1> INSIGHT 12:</h1>
<p>This shows that accidents involving more vehicles tend to have slightly more casualties, but the relationship is weak.
</p>

<h1> 13. Do different districts show different casualty numbers? </h1>

In [23]:
f_oneway(*[g["Number_of_Casualties"].values for _, g in uk_accidents.groupby("District Area")])

F_onewayResult(statistic=np.float64(10.307782461155597), pvalue=np.float64(0.0))

<h1>INSIGHT 13: </h1>
<p>This shows that some districts have more casualties per accident than others.</p>

<h1>DESCRIPTIVE COMPARISON </h1>

<h1>14. Which vehicle type is linked to the highest average severity?</h1>

In [24]:
uk_accidents.groupby("Vehicle_Type")["Accident_Severity"].mean().sort_values(ascending=False)

Vehicle_Type
Data missing or out of range             3.000000
Ridden horse                             3.000000
Motorcycle over 125cc and up to 500cc    2.846307
Goods over 3.5t. and under 7.5t          2.845965
Motorcycle over 500cc                    2.844102
Bus or coach (17 or more pass seats)     2.843330
Taxi/Private hire car                    2.843163
Car                                      2.840504
Motorcycle 125cc and under               2.840175
Van / Goods 3.5 tonnes mgw or under      2.836188
Goods 7.5 tonnes mgw and over            2.836145
Motorcycle 50cc and under                2.835406
Other vehicle                            2.833628
Agricultural vehicle                     2.833542
Minibus (8 - 16 passenger seats)         2.829517
Pedal cycle                              2.750000
Name: Accident_Severity, dtype: float64

<h1>INSIGHT 14: </h1>
<p>This shows that ridden horses and missing data cases have the highest severity. This suggests that motorcycles and mid-sized goods vehicles are linked to more severe accidents, while pedal cycles are the least severe.
</p>

<h1> 15. Which road surface leads to the highest casualties?</h1>

In [25]:
uk_accidents.groupby("Road_Surface_Conditions")["Number_of_Casualties"].mean().sort_values(ascending=False)

Road_Surface_Conditions
Flood over 3cm. deep    1.495098
Wet or damp             1.402693
Snow                    1.343997
Frost or ice            1.334603
Dry                     1.331892
Name: Number_of_Casualties, dtype: float64

<h1>INSIGHT 15:</h1>
<p>This shows that roads with flood over 3cm deep have the highest average casualties. This indicates that wet, snowy, or icy conditions also increase casualties, while dry roads have the lowest.</p>

<h1>16. What is the trend of average vehicles per accident across different years?</h1>

In [26]:
vehicle_year = uk_accidents.groupby(uk_accidents["Accident Date"].dt.year)["Number_of_Vehicles"].mean()
vehicle_year

Accident Date
2019    1.837784
2020    1.827145
2021    1.823475
2022    1.832004
Name: Number_of_Vehicles, dtype: float64

<h1>INSIGHT 16: </h1>
<p>This shows that the average number of vehicles per accident has remained fairly stable over time, ranging between 1.82 and 1.84, with only slight fluctuations across the years.</p>

<h1>17. Which accidents involved both motorcycles and high-speed roads (dual carriageways)?</h1>

In [27]:
motor_highspeed = uk_accidents[
    (uk_accidents["Vehicle_Type"].str.contains("Motorcycle")) &
    (uk_accidents["Road_Type"] == "Dual carriageway")
]

motor_summary = motor_highspeed["Vehicle_Type"].value_counts().sort_values(ascending=False)
motor_summary


Vehicle_Type
Motorcycle over 500cc                    1602
Motorcycle 125cc and under                877
Motorcycle over 125cc and up to 500cc     481
Motorcycle 50cc and under                 447
Data missing or out of range                0
Agricultural vehicle                        0
Car                                         0
Bus or coach (17 or more pass seats)        0
Minibus (8 - 16 passenger seats)            0
Goods over 3.5t. and under 7.5t             0
Goods 7.5 tonnes mgw and over               0
Other vehicle                               0
Pedal cycle                                 0
Ridden horse                                0
Taxi/Private hire car                       0
Van / Goods 3.5 tonnes mgw or under         0
Name: count, dtype: int64

<h1>INSIGHT 17: </h1>
<p>This shows that motorcycles on dual carriageways are involved in accidents, with Motorcycle over 500cc having the highest count, indicating that high-speed roads are riskier for powerful motorcycles.</p>

<h1>18. Are there vehicle types that rarely appear in accidents?</h1>

In [28]:
rare = uk_accidents["Vehicle_Type"].value_counts().nsmallest(5)
rare

Vehicle_Type
Ridden horse                          2
Data missing or out of range          4
Pedal cycle                          92
Minibus (8 - 16 passenger seats)    786
Agricultural vehicle                799
Name: count, dtype: int64

<h1>INSIGHT 18: </h1>
<p>This shows that agricultural vehicles and pedal cycle are involved in fewer accidents, indicating these vehicles are less common on accident-prone roads.</p>

<h1> 19. Which light condition occurs least often in accidents?</h1>

In [29]:
uk_accidents["Light_Conditions"].value_counts().nsmallest(1)

Light_Conditions
Darkness - lights unlit    1019
Name: count, dtype: int64

<h1>INSIGHT 19: </h1>
<p>This shows that darkness with lights unlit is the least common accident scenario, suggesting most accidents happen under some level of visibility.</p>

<h1>20. Which road types have the most accidents at night?</h1>

In [30]:
uk_accidents[uk_accidents["Light_Conditions"].str.contains("Darkness")]["Road_Type"].value_counts()

Road_Type
Single carriageway    51718
Dual carriageway      11679
Roundabout             4453
One way street         1459
Slip road               727
Name: count, dtype: int64

<h1>INSIGHT 20: </h1>
<p>This shows that single carriageways have the most accidents at night, followed by dual carriageways and roundabouts, indicating that these roads are more prone to nighttime accidents.</p>